juniper/src/macros/common.rs

```diff
@@ -0,0 +1,627 @@
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __juniper_impl_trait {
+    (
+        impl< < DefaultScalarValue > $(, $other: tt)* > $impl_trait:tt for $name:ty {
+            $($body:tt)+
+        }
+    ) => {
+        impl<$($other,)*> $crate::$impl_trait<$crate::DefaultScalarValue> for $name {
+            $($body)+
+        }
+    };
+    (
+        impl< <$generic:tt $(: $bound: tt)*> $(, $other: tt)* > $impl_trait:tt for $name:ty {
+            $($body:tt)+
+        }
+    ) => {
+       impl<$($other,)* $generic $(: $bound)*> $crate::$impl_trait<$generic> for $name
+        where
+            $generic: $crate::ScalarValue,
+            for<'__b> &'__b $generic: $crate::ScalarRefValue<'__b>,
+       {
+           $($body)+
+       }
+    };
+    (
+        impl<$scalar:ty $(, $other: tt )*> $impl_trait:tt for $name:ty {
+            $($body:tt)+
+        }
+    ) => {
+        impl<$($other, )*> $crate::$impl_trait<$scalar> for $name {
+            $($body)+
+        }
+    };
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __juniper_insert_generic {
+    (<DefaultScalarValue>) => {$crate::DefaultScalarValue};
+    (
+        <$generic:tt $(: $bound: tt)*>
+    ) => {
+        $generic
+    };
+    (
+        $scalar: ty
+    ) => {
+        $scalar
+    };
+}
+
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __juniper_parse_object_header {
+    (
+        callback = $callback:ident,
+        rest = <$($lifetime:tt),*> $name: ty $(: $ctxt: ty)* as $outname: tt
+        where Scalar = <$generic:tt $(: $bound:tt)*> $(| &$mainself:ident |)* {
```

Comment:
I'd still like to make the `where Scalar` clause optional in macros and default insert a `where Scalar = <S>` when omitted. This will ease upgrade pain and means every `graphql_object!` call doesn't need to be updated, only those with custom scalars.

Reply:
There is already [a default case](https://github.com/graphql-rust/juniper/pull/251/files/aa357f7935d0a361f17ded2ae5c69d64839f7c3e#diff-b5a47bf0125442e4083847d3907d57b5R101) without `where Scalar` clause. It defaults to using `DefaultScalarValue` because otherwise `FieldResult` and such things wouldn't work. (If we would default to a generic type `S` each field returning a result could not just return `FieldResult<T>` as in juniper <=0.10, but needs to return `FieldResult<T, S>` with some magic type `S`.)

Reply:
Oh, I missed that....the macros are huge 😄. Let's change the integrations to omit the clause then (example: https://github.com/graphql-rust/juniper/pull/251/files/aa357f7935d0a361f17ded2ae5c69d64839f7c3e#diff-e1dde797a5a7df30690ff359d95079d0R25)